### PR TITLE
[IGA] Coupling Nitsche Condition 

### PR DESCRIPTION
**Pull request by rickyaristio**
**Description**
This PR adds the new Nitsche-type coupling condition based on the work of @andreas-apostolatos on https://doi.org/10.1016/j.cma.2019.01.023. 

The test will be provided similarly to #8732.

### CODE REVIEW

File: applications/IgaApplication/custom_conditions/coupling_nitsche_condition.cpp [hunk 1]
```diff
@@ -0,0 +1,1440 @@
...
+        const auto& r_geometry_slave = GetGeometry().GetGeometryPart(1);
+        const Properties& r_properties = GetProperties();
+        const auto& r_N_master = r_geometry_master.ShapeFunctionsValues();
+        const auto& r_N_slave = r_geometry_slave.ShapeFunctionsValues();
+
+        const SizeType r_number_of_integration_points_master = r_geometry_master.IntegrationPointsNumber();
+        const SizeType r_number_of_integration_points_slave = r_geometry_slave.IntegrationPointsNumber();
+
+        //Constitutive Law initialisation
+        if (mConstitutiveLawVector_master.size() != r_number_of_integration_points_master)
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_conditions/coupling_nitsche_condition.cpp]: I would recommend just using the constitutive law vector and afterwards only checking the specific mast or slave one. In that way you need a little bit less storage.

File: applications/IgaApplication/custom_python/add_custom_strategies_to_python.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Riccardo Rossi
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_python/add_custom_strategies_to_python.h]: check the authors. I always neglect the authors completely, but that is up to you.

File: applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+///@name Type Definitions
+///@{
+
+///@}
+
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp:43]: please avoid unnecessary comments. That is only non-needed line overhead.

File: applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    /// Constructor.
+    EigensolverNitscheStabilizationScheme() : Scheme<TSparseSpace,TDenseSpace>() {}
+
+    /// Destructor.
+    ~EigensolverNitscheStabilizationScheme() override {}
+
+    ///@}
+    ///@name Operators
+    ///@{
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp]: remove those blocks where not needed

File: applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        { 
+            rCurrentElement.CalculateLocalSystem(LHS_Contribution,RHS_Contribution,CurrentProcessInfo); 
+        }
+
+        rCurrentElement.EquationIdVector(EquationId,CurrentProcessInfo);
+
+        KRATOS_CATCH("")
+    }
+
+    void CalculateSystemContributions(
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp:110]: give a comment as description why you need to override this function.

File: applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void CalculateSystemContributions(
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_schemes/eigensolver_nitsche_stabilization_scheme.hpp:86]: comments would be nice.

File: applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+    };
+
+    BuilderAndSolverPointerType& pGetBuilderAndSolver()
+    {
+        return mpBuilderAndSolver;
+    };
+
+    SparseMatrixType& GetStiffnessMatrix()
+    {
+//         if (mpStiffnessMatrix == nullptr)
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp]: You can remove that commented code

File: applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * Performs all the required operations that should be done (for each step)
+     * before solving the solution step.
+     * A member variable should be used as a flag to make sure this function is called only once per step.
+     */
+    void InitializeSolutionStep() override
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp:275]: Please highleight specifically what you do here. Soon you will be happy about your comments.

File: applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
```
Comment (tteschemacher) [inline applications/IgaApplication/custom_strategies/custom_strategies/eigensolver_nitsche_stabilization_strategy.hpp]:
```suggestion
```
remove all that

File: applications/IgaApplication/python_scripts/nitsche_stabilization_process.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        # Compute the eigen values
+        eigen_linear_solver = eigen_solver_factory.ConstructSolver(self.params["eigen_system_settings"])
+        builder_and_solver = KratosMultiphysics.ResidualBasedBlockBuilderAndSolver(eigen_linear_solver)
+        eigen_scheme = IGA.EigensolverNitscheStabilizationScheme()
+        eigen_solver = IGA.EigensolverNitscheStabilizationStrategy(self.model_part, eigen_scheme, builder_and_solver)
+        eigen_solver.Solve()
+
+        # Compute the Nitsche stabilization factor
+        eigenvalue_nitsche_stabilization_vector = current_process_info.GetValue(IGA.EIGENVALUE_NITSCHE_STABILIZATION_VECTOR)
+        nitsche_stabilization_factor= eigenvalue_nitsche_stabilization_vector[eigenvalue_nitsche_stabilization_vector.Size()-1]*4*self.params["number_of_couplings"].GetInt()
```
Comment (tteschemacher) [inline applications/IgaApplication/python_scripts/nitsche_stabilization_process.py:81]: we should check if the stabilization factor is really linearly dependent on the number of couplings. That sound a bit weird to me...

**Review by tteschemacher — COMMENTED** [review]
Hey @rickyaristio,

thank you very much for this really great development. I have some minor comments and then I will be more than happy to merge as soon as possible.

File: applications/IgaApplication/python_scripts/nitsche_stabilization_process.py (hunk 10 above)
Comment (rickyaristio) [reply applications/IgaApplication/python_scripts/nitsche_stabilization_process.py:81]: Noted.

File: applications/IgaApplication/custom_conditions/coupling_nitsche_condition.cpp (hunk 1 above)
Comment (rickyaristio) [reply applications/IgaApplication/custom_conditions/coupling_nitsche_condition.cpp]: There may be a case where the constitutive law vectors may be different for both master and slave. Well, I open to discuss this in detail.

Comment (rickyaristio) [reply applications/IgaApplication/custom_conditions/coupling_nitsche_condition.cpp]: > There may be a case where the constitutive law vectors may be different for both master and slave. Well, I open to discuss this in detail.

Sorry for the misunderstanding. I've changed how the constitutive law vector is assigned in the recent commit.

Comment (tteschemacher) [reply applications/IgaApplication/custom_conditions/coupling_nitsche_condition.cpp]: No worries, thanks a lot for the change!

**Review by tteschemacher — APPROVED** [review]
Thank you very much @rickyaristio for the really comprehensive and complex implementation!

**Pull request merged**